### ISSUE TITLE
DNS Challenge Verification and Integration Tests

### ISSUE THREAD

**Issue by cristiangraz** [xenolf/lego#259]
I'm using Lego with a custom DNS `acme.ChallengeProvider` and writing integration tests for our SSL integration. When using the DNS Challenge, Lego checks `acme.RecursiveNameServers` for the TXT records before notifying Boulder that the challenge is ready.

While this is a nice feature, it poses challenges with integration tests. Boulder has a test DNS server that lets you `POST` TXT records to them so they can "verify" the challenge in a test environment. Right now, we need to create a fake DNS server and point `acme.RecursiveNameServers` to it in order to be able to run our tests while using Lego. This causes a lot of extra code and possible errors in order to work around this additional check.

I was curious if anyone had any thoughts or suggestions on how they've addressed this. You can currently override how the timeout works with a `ChallengeProviderTimeout`, but you can't disable the check all together.

I was thinking that in cases like ours where a user is using their own DNS `acme.ChallengeProvider`, they can nil out `acme.RecursiveNameservers` and we could add this to the top of `checkDNSPropagation` function:

```go
func checkDNSPropagation(fqdn, value string) (bool, error) {
	// If name servers are empty, always return true.
	if len(RecursiveNameservers) == 0 {
		return true, nil
	}

	// ...
}
```
We could also implement a new exported bool variable to disable this check. Let me know your thoughts, I'm happy to submit a PR but wanted to open up for discussion first in case someone else has already found a simpler way around this.

https://github.com/xenolf/lego/blob/master/acme/dns_challenge.go#L74-L91

**Comment by xenolf** [comment]
One thing that comes to mind is for lego to export the `preCheckDNS` variable. By doing this you could supply your own pre-check function and do whatever you need to do. We are using this in a test in order to get around the DNS check :)

**Comment by cristiangraz** [comment]
Ahh I saw that earlier but the idea of exporting hadn't occurred to me. That would work great, I'll start on a PR shortly!

**Comment by xenolf** [comment]
Closed via #260 
@cristiangraz Thanks for implementing this! :)
